### ISSUE TITLE
Allow syncing to local filesystem / new sync target

### ISSUE THREAD

**Issue by markokocic** [200ok-ch/organice#730]
**Is your feature request related to a problem? Please describe.**
Right now it is not possible to configure organice to use local file system to sync / store org files. Only certain cloud providers are supported. For users that are already syncing their org files through different (server, desktop and mobile) devices (manually, using git, syncthing, rsync, network shares, ...) requiring another sync system setup to use organice is an overkill.

**Describe the solution you'd like**
I'd like to be able to configure organice to use a local folder (with subfolders) on the device I use to read from and store org files. It should also be able to automatically detect external changes and has conflict resolution a same way as for current synchronization targets.

**Describe alternatives you've considered**
As I want to avoid using cloud services to store my private data, and setting up, maintaining and securing a personal WebDAV server is too much of a hassle, I don't see any alternative.

**Additional context**
Adding support to sync to local filesystem would decouple syncing from the app, and allow users to use whatever external sync mechanism they like.

**Comment by schoettl** [comment]
See also #93

**Comment by munen** [comment]
@markokocic Which JS API do you propose to employ for accessing the local file system?

**Comment by markokocic** [comment]
I was under impression that progressive web apps have local filesystem access possibility, but looks like I was wrong :(

That being said, [IPFS ](https://ipfs.io/) as a concept for data sharing looks interesting, not sure if there's interest in organice.

**Comment by fullofcaffeine** [comment]
+1 would love to be able to use organic with syncthing.

**Comment by munen** [comment]
@fullofcaffeine I haven’t used Syncthing before and have no experience with it. From a quick glance, I am not sure if it technically can be used from a browser based app. However, if it is: Synchronization in organice is built using the strategy pattern which is consumed here[1]. Implementing a new back-end implies implementing 8 functions[2]. For Dropbox, it’s about 120 LOC JavaScript. 

Would you be interested to integrate a new back-end? I’ll be happy to help with questions, code review and testing. 

https://github.com/200ok-ch/organice/blob/master/src/actions/sync_backend.js
https://github.com/200ok-ch/organice/blob/master/src/sync_backend_clients/dropbox_sync_backend_client.js#L134-L141

**Comment by munen** [comment]
@markokocic @fullofcaffeine @non-jedi Do you plan to work on this?

Otherwise I'll have to close this for one month of inactivity due to https://organice.200ok.ch/documentation.html#how_we_work_with_issues

**Comment by mrloop** [comment]
Looks like some file access is possible with the non standard [File System Access API](https://caniuse.com/?search=File%20System%20Access%20API) but only on some desktop browsers. I also see there's a ponyfill here https://github.com/GoogleChromeLabs/browser-fs-access that may work. I'll try and find time and put a proof of concept together with this is see if it'll work.

**Comment by munen** [comment]
@mrloop Sweet, looking forward to seeing a PoC. Thank you for stepping up :pray:

**Comment by mrloop** [comment]
A work in progress of the POC can be found here https://github.com/mrloop/organice/pull/1 it has not got very far.

The purpose of the of the POC was to examine if it was possible to create a file system adapter for organice using the non standard [File System Access API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API). The File System Access API only works on some desktop browsers. As organice is mobile focused using a polyfill/ponyfill is needed.

## Libararies

Initially looked at [browser-fs-access](https://github.com/GoogleChromeLabs/browser-fs-access). It does not expose the File System Access API, but rather provides its own more limited API that uses the File System Access API or if not available fallsback to using `<input type="file">` and `<a download>`. It does not support the File System Access API `showDirectoryPicker` and `FilesSystemDirectoryHandle` meaning browser-fs-access could not be used to nicely integrate with the organice file browser. It may be possible to use it to open and edit individual files but would feel out of place in terms of UX. Might be worth a further look at, one concern is the fallback may result in lots of open / save dialogs opening up when organice want to read write files.

An alternative library is [native-file-system-adapter](https://github.com/jimmywarting/native-file-system-adapter/) which has an API which matches the File System Access spec meaning it could integrate well with the organice file browser. 

## Issues

However I ran into a few issues while trying to use the native-file-system-adapter. create-react-app intentionally prevent transpiling 3rd party dependencies that rely on experimental transforms, see this issues https://github.com/facebook/create-react-app/issues/4648 . native-file-system-adapter uses experimental transforms and does not supply a transpiled version

```sh
Failed to compile.

./node_modules/native-file-system-adapter/src/FileSystemDirectoryHandle.js
SyntaxError: /home/ewan/open-source/organice/node_modules/native-file-system-adapter/src/FileSystemDirec$

   7 |   [kAdapter]
   8 |
>  9 |   constructor (adapter) {
     |   ^
  10 |     super(adapter)
  11 |     this[kAdapter] = adapter
  12 |   }

Add @babel/plugin-proposal-class-properties (https://git.io/vb4SL) to the 'plugins' section of your Babe$
```

I tried compiling native-file-system-adapter with `esbuild`

```
npx esbuild src/es6.js --outfile=./dist/native-file-system-adapter.js --bundle --format=esm --target=es2020
```

However it contains top-level await TLA, and only supported for a target of esnext and not for something that will run in a browser, for instance es2020. Can't be bundled and transpiled with esbuild

https://esbuild.github.io/content-types/#javascript
https://github.com/evanw/esbuild/issues/253

Alternatively Webpack has support for TLA and can be used with babel to transpile the class properties. I've tried a few different configuration of Webpack but have been unable to get a working bundle. I have the same concern for this library as I do for browser-fs-access, that the fallback may result in lots of open / save dialogs opening up when organice wants to read and write files. I'd hoped that a POC would show if this was an issue and how bad it was for UX. 

I've lost any enthusiasm to pursue this further but maybe my notes will help somebody else looking at local file system support.

**Comment by munen** [comment]
Hi @mrloop 

Thank you for checking out this issue and providing detailed information. This will indeed be valuable for the next person interested in this topic.

I'm sorry this didn't lead to where you wanted to go.

Thank you, again and kind regards :pray: :bow:

**Comment by munen** [comment]
Closing as this cannot be implemented as previously stated and has now been proven by an independent POC.

**Comment by yantar92** [comment]
Rather than trying to access local FS direcly, may it be possible to create a simple local server that will serve the connector between organice and the file system?

**Comment by munen** [comment]
@yantar92 Absolutely! You can always set up a local WebDAV server. organice ships one for testing in a Docker configuration, but there are many options for WebDAV servers. Here's some documentation regarding organice and WebDAV:

- https://organice.200ok.ch/documentation.html#orge0ee6c2
- https://organice.200ok.ch/documentation.html#faq_webdav

**Comment by yantar92** [comment]
"Zen Monk Alain M. Lafon" ***@***.***> writes:

> @yantar92 Absolutely! You can always set up a local WebDAV server. organice ships one for testing in a Docker configuration, but there are many options for WebDAV servers. Here's some documentation regarding organice and WebDAV:

Well. I am not up to setting up the whole ngnix/apache. And rclone
webdav server is not supported even if I use no authentication (rclone
does not provide CORS).

Also, I am wondering if there is any webdav server for android. That
would work nicely: Organice > WebDav > Syncthing.

-- 
Ihor Radchenko // yantar92,
Org mode contributor,
Learn more about Org mode at <https://orgmode.org/>.
Support Org development at <https://liberapay.com/org-mode>,
or support my work at <https://liberapay.com/yantar92>

**Comment by munen** [comment]
Yes, there are users running organice and webdav locally on their phone. Here's an example: https://github.com/200ok-ch/organice/tree/master/contrib/organice-pinephone

**Comment by yantar92** [comment]
"Zen Monk Alain M. Lafon" ***@***.***> writes:

> Yes, there are users running organice and webdav locally on their phone. Here's an example: https://github.com/200ok-ch/organice/tree/master/contrib/organice-pinephone

Thanks for the link, but I am afraid that it is not very useful for
Android phones; even rooted. Pinephone is essentially a Linux-driven
phone and the instructions assume a proper Linux.

-- 
Ihor Radchenko // yantar92,
Org mode contributor,
Learn more about Org mode at <https://orgmode.org/>.
Support Org development at <https://liberapay.com/org-mode>,
or support my work at <https://liberapay.com/yantar92>

**Comment by munen** [comment]
@yantar92 Good news, then. I just added another piece of contrib documentation of how to set up organice on Android with a local WebDAV server: https://github.com/200ok-ch/organice/tree/master/contrib/organice-on-android-with-webdav-server

**Comment by yantar92** [comment]
"Zen Monk Alain M. Lafon" ***@***.***> writes:

> @yantar92 Good news, then. I just added another piece of contrib documentation of how to set up organice on Android with a local WebDAV server: https://github.com/200ok-ch/organice/tree/master/contrib/organice-on-android-with-webdav-server

Thanks!

This looks useful, but hard to follow:

"Note that as of 2020-08-28, the required version of Node.js for
compiling organice is v12.13.1, however, the nodejs-lts package is on
v12.18.3"

This statement still holds true today and I was unable to compile
Organice locally in termux. The provided link
https://wiki.termux.com/wiki/Building_packages requires a lot of
research to follow further as it is just a generic guide not specific to
installing the required Node.js version.

"I recommend these settings in your config: http://ix.io/2vuX"

Webdav settings are awkwardly placed to third-party service and are a
subject of removal. Also, I was unable to access the WebDav server from
mobile browser. Something may still be missing in the guide.

-- 
Ihor Radchenko // yantar92,
Org mode contributor,
Learn more about Org mode at <https://orgmode.org/>.
Support Org development at <https://liberapay.com/org-mode>,
or support my work at <https://liberapay.com/yantar92>

**Comment by munen** [comment]
For future reference, I have added an ADR on how to implement a new synchronization back-end: https://organice.200ok.ch/documentation.html#adr-003
